Issue #141: Support device back button for navigation

csadilek opened the pull request:
This works out quite OK with our existing component functionality. So, I think, right now this seems enough?

One thought I had is that we might want to address the train wrecks by referencing the individual components directly, but that's a separate discussion for later (when we add it to Fenix).

csadilek commented:
Hold on. There's a piece missing which I thought I had already added, but didn't.

Re: toolbar and other components that are interested in the back navigation happening. My thoughts here are the the toolbar presenter (or others) will react to a navigate event (rather than the UI specific back pressed event). However, we do have a NavigateStateChange event (whether or not back and forward are possible), but no Navigation event (when the user actually goes back and forward). The latter is still needed.

Working on adding this here...

pocmo commented:
> Re: toolbar and other components that are interested in the back navigation happening. My thoughts here are that the toolbar presenter (and/or other presenters) will react to a navigation event (rather than the UI specific back pressed event)

Oh, I was a bit unclear about that: The toolbar is not interested in the navigation events but instead the actual back button press:
* Open Fennec
* Edit the URL
* Close the keyboard
* Press back

-> Fennec will switch from "editing mode" to "display mode".

In fact when looking at Fennec there are multiple components that may want to handle the back button press themselves in a very specific order.
https://dxr.mozilla.org/mozilla-central/source/mobile/android/base/java/org/mozilla/gecko/BrowserApp.java#1203

In my toolbar branch I added something like `onBackPressed: Boolean` that the app could call.

Anyhow, that's something to address later.

csadilek commented:
@pocmo I had to reset my stack a couple of times here because I wasn't happy with the abstractions I came up with i.e. I added more complexity than value. I've added a commit now that makes navigation events observable, and sticks to our current architecture. At least for this part, I like the implementation, as it allows any presenter to observe these events because they are already session observers anyway.

On the UI events (e.g. device back press): So far, I haven't been able to think of an abstraction that really makes this easier. We'd need a component that observes `onBackPressed` and acts on it by invoking handlers that can be registered by the app or other components. This is basically an abstraction without a benefit as the app and component can already do this. If we could define the exact default behaviour of `onBackPressed`, then of course there would be a benefit of having that reusable code, but that component would then need to live in a separate module with references to toolbar etc. I will think some more about this.

Hope that makes sense. Wdyt? Can you take a look at the commit I added so we can decide if we want to include it?

sputnikci commented on components/browser/engine-gecko/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineSessionTest.kt:
```diff
@@ -96,28 +89,46 @@ class GeckoEngineSessionTest {
+                BundleEventListener { _, _, _ -> geckoEventReceived = true }, "GeckoView:GoBack")
+
+        var sessionEventReceived = false
+        engineSession.register(object: EngineSessionObserver() {
```
\[ktlint\] WARNING: \[colon\-spacing\] Missing spacing before ":" in column 38

sputnikci commented on components/browser/engine-gecko/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineSessionTest.kt:
```diff
@@ -96,28 +89,46 @@ class GeckoEngineSessionTest {
+            BundleEventListener { _, _, _ -> geckoEventReceived = true }, "GeckoView:GoForward")
+
+        var sessionEventReceived = false
+        engineSession.register(object: EngineSessionObserver() {
```
\[ktlint\] WARNING: \[colon\-spacing\] Missing spacing before ":" in column 38

sputnikci commented on components/browser/engine-gecko/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineSessionTest.kt:
```diff
@@ -96,28 +89,46 @@ class GeckoEngineSessionTest {
+        assertTrue(sessionEventReceived)
+    }
+
+    open class EngineSessionObserver: EngineSession.Observer {
```
\[ktlint\] WARNING: \[colon\-spacing\] Missing spacing before ":" in column 37

sputnikci commented on components/browser/engine-system/src/test/java/mozilla/components/browser/engine/system/SystemEngineSessionTest.kt:
```diff
@@ -22,11 +23,8 @@ class SystemEngineSessionTest {
-        engineSession.register(object : EngineSession.Observer {
-            override fun onLoadingStateChange(loading: Boolean) { }
-            override fun onLocationChange(url: String) { }
+        engineSession.register(object: EngineSessionObserver() {
```
\[ktlint\] WARNING: \[colon\-spacing\] Missing spacing before ":" in column 38

sputnikci commented on components/browser/engine-system/src/test/java/mozilla/components/browser/engine/system/SystemEngineSessionTest.kt:
```diff
@@ -49,8 +47,14 @@ class SystemEngineSessionTest {
         `when`(engineSession.currentView()).thenReturn(webView)
 
+        var sessionEventReceived = false
+        engineSession.register(object: EngineSessionObserver() {
```
\[ktlint\] WARNING: \[colon\-spacing\] Missing spacing before ":" in column 38

sputnikci commented on components/browser/engine-system/src/test/java/mozilla/components/browser/engine/system/SystemEngineSessionTest.kt:
```diff
@@ -59,7 +63,22 @@ class SystemEngineSessionTest {
         `when`(engineSession.currentView()).thenReturn(webView)
 
+        var sessionEventReceived = false
+        engineSession.register(object: EngineSessionObserver() {
```
\[ktlint\] WARNING: \[colon\-spacing\] Missing spacing before ":" in column 38

sputnikci commented on components/browser/engine-system/src/test/java/mozilla/components/browser/engine/system/SystemEngineSessionTest.kt:
```diff
@@ -59,7 +63,22 @@ class SystemEngineSessionTest {
+        assertTrue(sessionEventReceived)
+    }
+
+    open class EngineSessionObserver: EngineSession.Observer {
```
\[ktlint\] WARNING: \[colon\-spacing\] Missing spacing before ":" in column 37

sputnikci commented on components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt:
```diff
@@ -4,6 +4,7 @@
 
 package mozilla.components.feature.toolbar
 
+import android.util.Log
```
\[ktlint\] WARNING: \[no\-unused\-imports\] Unused import in column 1

csadilek commented:
@pocmo OK, had another idea, which I added as a separate commit. This cleans up the app logic by dispatching the onBackPressed logic to the specific features. They can then handle it whichever way they like i.e. the SessionFeature goes back in history if possible, the ToolbarFeature makes the presenter switch the Toolbar into display mode.

pocmo commented:
Mh. I don't think c38f5fde8ad984bfaedf3ec23397f762c81b39a3 will be needed. I can't think of any component that will actually need to observe Navigation events. So far every component in Focus is interested in state changes only. For navigation that's usually changes in the URL etc.

pocmo commented:
571a67207ae4deeee5f0762879b149b693fff965 looks great and passing back presses to the feature is a nice way to not needing to route this to a specific component.

pocmo commented on samples/browser/src/main/java/org/mozilla/samples/browser/MainActivity.kt:
```diff
@@ -48,6 +48,15 @@ class MainActivity : AppCompatActivity() {
+            super.onBackPressed()
+        }
+    }
+
```
This will need to be ordered and exit early I think:

1. Toolbar
2. Session
3. (parent) Activity

If a previous component handled the back press then it shouldn't be forwarded to the next one.

csadilek replied:
Yes, that makes sense. I was trying to simplify in a sense that eventually we'd loop over all features and invoke their `handleBackPressed`, where order doesn't matter. If none handled the event, we'd call `super.`, but in either case, I think that will be in the apps control. So, fine with changing it here.

csadilek commented:
OK, I'll stash c38f5fd for now. Re: 571a672: Cool, I think it's a good compromise and cleans up the app logic a bit. I do think we might want to introduce an interface for `Feature` eventually but we don't have to do it now.

csadilek commented:
@pocmo OK, this is all updated now.

pocmo commented:
👍
